### ISSUE TITLE
context and merge

### ISSUE THREAD

**Issue by galindro** [Praqma/helmsman#578]
I would like to do the following, but I'm hitting this limitation: `When merging multiple DSFs, context from the firs DSF in the list gets overridden by the context in the last DSF.`. I'm opening this issue to confirm if contexts really doesn't works with merge feature and if it doesn't, why is it documented in https://github.com/Praqma/helmsman/blob/master/docs/how_to/misc/merge_desired_state_files.md? I would like to know if it is too hard to make it work with merge feature.

I have these 3 DSF files:

### helmsman-common.yaml
```bash
helmRepos:
  bitnami: 'https://charts.bitnami.com/bitnami'

settings:
  reverseDelete: true

namespaces:
  external-dns:
    protected: false
```
### helmsman-private.yaml
```bash
context: ict-private

appsTemplates:
  private-default: &private-default
    enabled: true
    wait: true
    timeout: 600

apps:
  external-dns-private:
    <<: *private-default
    namespace: "external-dns"
    chart: "bitnami/external-dns"
    version: "4.4.1"
    valuesFiles:
      - "./common/ict-tolerations.yaml"
      - "./common/ict-private-node-affinity.yaml"
      - "external-dns/values-private.yaml"
    priority: -600
```

### helmsman-public.yaml
```bash
context: ict-public

appsTemplates:
  public-default: &public-default
    enabled: true
    wait: true
    timeout: 600

apps:
  external-dns-public:
    <<: *public-default
    namespace: "external-dns"
    chart: "bitnami/external-dns"
    version: "4.4.1"
    valuesFiles:
      - "./common/ict-tolerations.yaml"
      - "./common/ict-private-node-affinity.yaml"
      - "external-dns/values-public.yaml"
    priority: -600
```

If I run `helmsman apply -subst-ssm-values -subst-env-values -show-diff -f helmsman-common.yaml -f helmsman-private.yaml`, it runs fine and install the application on the correct context: ict-private.

The same occurs when I run `helmsman apply -subst-ssm-values -subst-env-values -show-diff -f helmsman-common.yaml -f helmsman-public.yaml`, but on the ict-public context.

However, if I run `helmsman apply -subst-ssm-values -subst-env-values -show-diff -f helmsman-common.yaml -f helmsman-private.yaml -f helmsman-public.yaml`, then I get this error:

`2021-01-21 13:02:16 CRITICAL: Release [ external-dns-public ] in namespace [ external-dns ] already exists but is not managed by the current context: [ ict-private ]. Applying changes will likely cause conflicts. Change the release name or namespace.`

**Comment by luisdavim** [comment]
Hi, I think this is the expected behaviour, the merge merges all the files into one if they have overlapping properties the last one will take precedence and will overwrite the previous ones, the idea is that you have a common base and then can set environment specific properties in environment specific additional files, if you have separate contexts for the different envs you'll need to run them separately, `helmsman-common.yaml` + `helmsman-private.yaml` and then `helmsman-common.yaml` + `helmsman-public.yaml`.

The merge is literally merging the state files we don't do any special parsing to check if the individual files had different contexts and I'm not sure we'd want to do that, it would add a significant extra amount of complexity but not that much extra value since you can achieve what you want by running the 2 envs from separate commands and some users might actually want to overwrite the context when merging DST files and that's what the document you've linked is referring to, maybe we need to reword it to make it more clear.

**Comment by galindro** [comment]
Yes I know that I can do it by issuing 2 commands. This is exactly the workaround that I implemented. 

And what about let the users decide what they want in the merge? Perhaps a command line flag (--keep-contexts). It is up to you to decide.

If it isn't feasible, at least the docs should be put in other place, as you said.

**Comment by luisdavim** [comment]
I agree we should make the docs more clear but I'm not sure if the extra complexity of allowing what you're asking is worth the extra value it would bring....

**Comment by galindro** [comment]
Well, in my case I administer multiple clusters. Some of them requires some packages that others don't and vice-versa. My CI pipeline is generic for all of them. So, to circumvent this situation, I created a bunch of shell scripting logic. With such proposed flag, this wouldn't be necessary. Anyway, the logic is already in place and it is working. So, it's up to you to decide.
